qe/drivers.py

```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        :param V_ik, W_ik: local working variables, numpy arrays
+
+        :return new values for dim_S, V_ik, and W_ik following implicit restart
+        """
+        if n_newvecs == 0:
+            V_inew = X_ik[:, :n_eig]
+        else:
+            V_inew = np.c_[X_ik[:, :n_eig], V_ik[:, -n_newvecs:]]
+        # Take leading n_eig Ritz vectors as new guess vectors
+        dim_S = V_inew.shape[1]  # New subspace dimension
+        V_ik = np.zeros((self.local_size, 0), dtype="float")
+        W_ik = np.zeros((self.local_size, 0), dtype="float")
+        # Initialize; normalize first basis vector
+        v_inew = np.array(V_inew[:, 0], dtype="float")
+        v_new = np.zeros(self.full_size, dtype="float")
+        self.comm.Allgatherv(
+            [v_inew, MPI.DOUBLE], [v_new, self.distribution, self.offsets, MPI.DOUBLE]
+        )
+        V_ik = np.c_[V_ik, v_inew / np.linalg.norm(v_new)]
+        for j in range(1, dim_S):
+            # Orthogonalize next vector against previous ones in restart basis
+            v_inew, norm_vnew = self.mgs(V_ik[:, :j], V_inew[:, j])
+            V_ik = np.c_[V_ik, v_inew]  # Update basis
+        n_newvecs = dim_S
+
+        return dim_S, n_newvecs, V_ik, W_ik
+
+    def mgs(self, V_ik, t_ik):
+        """Parallel implementation of Modified Graham-Schmidt (MGS).
+        Takes piece of new guess vector t_ik, and orthogonalizes it against trial subspace V_k.
+
+        :param V_ik: local work variable, numpy matrix
+        :param t_ik: local work variable, numpy vector
+
+        :return orthonormalized vector t_ik
+        """
+        for j in range(V_ik.shape[1]):  # Iterate through k basis vectors
+            c_j = np.copy(np.inner(V_ik[:, j], t_ik))  # Each process computes partial inner-product
+            self.comm.Allreduce([c_j, MPI.DOUBLE], [c_j, MPI.DOUBLE])  # Default op=SUM
+            t_ik = t_ik - c_j * V_ik[:, j]  # Remove component of t_ik in V_ik
+        t_k = np.zeros(
+            self.full_size, dtype="float"
+        )  # Pre-allocate space to receive new guess vector
+        self.comm.Allgatherv([t_ik, MPI.DOUBLE], [t_k, self.distribution, self.offsets, MPI.DOUBLE])
+        norm_tk = np.linalg.norm(t_k)
+        return t_ik / norm_tk, norm_tk  # Return new orthonormalized vector
+
+    def preconditioning(self, D_i, lambda_k, r_ik):
+        """Preconditon next guess vector
+
+        :param D_i: diagonal portion of local Hamiltonian, as a numpy vector
+        :param lambda_k: an eigenvalue, as a scalar
+        :param r_ik: residual, a numpy vector
+
+        :return numpy vector
+        """
+        M_k = np.diag(
+            np.clip(np.reciprocal(D_i - lambda_k), a_min=-1e5, a_max=1e5)
+        )  # Build diagonal preconditioner
+        return np.dot(M_k, r_ik)
+
+    def distributed_davidson(self, H_i, n_eig, n_guess, eps, max_iter, q, driven_by="explicit"):
+        """Davidson's method implemented in parallel. The Hamiltonian
```

Comment:
We can have only one davidson, don't need to save the other one

Reply:
Done!